## Задание 1

* Импортируйте библиотеки pandas и numpy.
* Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
* Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
* составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
* Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
* Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
* Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [2]:
import numpy as np
import pandas as pd

from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [4]:
# import warnings

# warnings.filterwarnings('ignore')

#### Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.

In [3]:
boston = load_boston()

In [4]:
X = pd.DataFrame(data=boston.data, columns=boston['feature_names'])

In [5]:
y = pd.DataFrame(data=boston.target, columns=['TARGET'])

#### Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

In [6]:
X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.3, random_state=42)

#### Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [7]:
lr = LinearRegression(copy_X=True)

#### Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [8]:
lr_ref = lr.fit(X_train, y_train)

In [9]:
y_pred = lr_ref.predict(X_test)

#### Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [10]:
r2_score(y_true=y_test, y_pred=y_pred)

0.7112260057484943

## Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble. сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

#### Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble. сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.

In [12]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(
    n_estimators=1000,
    max_depth=12,
    random_state=42
)

#### Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.¶

In [13]:
X, y = load_boston(return_X_y=True)
X = pd.DataFrame(data=X)
y = pd.DataFrame(data=y)

In [14]:
X_train, X_test, y_train, y_test =\
    train_test_split(X, y, test_size=.3, random_state=42)

len(X_test) / len(X), len(y_test) / len(y)

(0.30039525691699603, 0.30039525691699603)

In [15]:
model_ref = model.fit(X_train, y_train.values[:, 0])

#### Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.

In [16]:
y_pred = model_ref.predict(X_test)
y_pred[:2], y_test.head(2)

(array([22.80641237, 31.13146352]),
         0
 173  23.6
 274  32.4)

In [17]:
rfr_r2 = r2_score(y_true=y_test, y_pred=y_pred.reshape(-1, 1))
rfr_r2

0.87472606157312

In [18]:
lr_r2 = 0.7112260057484974

In [19]:
r2_scale = round(100 * (rfr_r2 - lr_r2) / rfr_r2, 2)
r2_scale

18.69

#### Вывод: модель типа RandomForestRegressor обучилась на данном датасете лучше (дала более достоверный прогноз) чем LinearRegression на 18.69 %

## Задание 3¶

Вызовите документацию для класса RandomForestRegressor, 
найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.

In [20]:
print(RandomForestRegressor.__doc__)


    A random forest regressor.

    A random forest is a meta estimator that fits a number of classifying
    decision trees on various sub-samples of the dataset and uses averaging
    to improve the predictive accuracy and control over-fitting.
    The sub-sample size is controlled with the `max_samples` parameter if
    `bootstrap=True` (default), otherwise the whole dataset is used to build
    each tree.

    Read more in the :ref:`User Guide <forest>`.

    Parameters
    ----------
    n_estimators : int, default=100
        The number of trees in the forest.

        .. versionchanged:: 0.22
           The default value of ``n_estimators`` changed from 10 to 100
           in 0.22.

    criterion : {"mse", "mae"}, default="mse"
        The function to measure the quality of a split. Supported criteria
        are "mse" for the mean squared error, which is equal to variance
        reduction as feature selection criterion, and "mae" for the mean
        absolute error.

       

#### найдите информацию об атрибуте feature_importances_.¶
    
    feature_importances_ : ndarray of shape (n_features,)
        The impurity-based feature importances.
        The higher, the more important the feature.
        The importance of a feature is computed as the (normalized)
        total reduction of the criterion brought by that feature.  It is also
        known as the Gini importance.

        Warning: impurity-based feature importances can be misleading for
        high cardinality features (many unique values). See
        :func:`sklearn.inspection.permutation_importance` as an alternative.

#### С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.

In [22]:
model = RandomForestRegressor(
    n_estimators=1000,
    max_depth=12,
    random_state=42
)

boston = load_boston()

X = pd.DataFrame(data=boston.data)
y = pd.DataFrame(data=boston.target)

X_train, X_test, y_train, y_test =\
    train_test_split(X, y, test_size=.3, random_state=42)

rt_fitted = model.fit(X_train, y_train.values[:, 0])

In [23]:
importances = rt_fitted.feature_importances_

In [24]:
importances.sum()

1.0

## Задание 4¶
В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию по библиотеке Matplotlib, это датасет Credit Card Fraud Detection.Для этого датасета мы будем решать задачу классификации - будем определять,какие из транзакциции по кредитной карте являются мошенническими.Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки),так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель.Мы будем вычислять AUC, то есть площадь под кривой ROC.

Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.

Загрузите датасет creditcard.csv и создайте датафрейм df.

С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка несбалансирована.Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков.Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма:

pd.options.display.max_columns = 100.
Просмотрите первые 10 строк датафрейма df.
Создайте датафрейм X из датафрейма df, исключив столбец Class.
Создайте объект Series под названием y из столбца Class.
Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: test_size=0.3, random_state=100, stratify=y.
У вас должны получиться объекты X_train, X_test, y_train и y_test.
Просмотрите информацию о их форме.
Для поиска по сетке параметров задайте такие параметры:
parameters = [{
  'n_estimators': [10, 15],
  'max_features': np.arange(3, 5),
  'max_depth': np.arange(4, 7)
}]
Создайте модель GridSearchCV со следующими аргументами:
estimator=RandomForestClassifier(random_state=100),
param_grid=parameters,
scoring='roc_auc',
cv=3.
Обучите модель на тренировочном наборе данных (может занять несколько минут).
Просмотрите параметры лучшей модели с помощью атрибута best_params_.
Предскажите вероятности классов с помощью полученнной модели и метода predict_proba.
Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba. Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных, используя в качестве аргументов массивы y_test и y_pred_proba.

In [25]:
import numpy as np
import pandas as pd
pd.options.display.max_columns = 100

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score

In [ ]:
DATASET_FILE = './creditcard.csv'

In [ ]:
df = pd.read_csv(SRC_DATASET_FILE)

In [ ]:
df.Class.value_counts(normalize=True)

In [ ]:
df.info()

In [ ]:
df.head(10)